docs/prover/techniques/index.md

```diff
@@ -1,31 +1,121 @@
+In this chapter, we describe some of the techniques used inside the Certora
+Prover. While this knowledge is not essential for using the Prover, it can
+sometimes be helpful when the Prover does not behave as expected, for instance
+in case of a timeout.
 
 (control-flow-splitting)=
-# Control Flow Splitting
+# Control flow splitting
 
-There is a brief explanation of control flow splitting in the 
+
+```{note}
+In addition to the text-form documentation below,  there is a brief explanation 
+of control flow splitting in the 
 [webinar on timeouts](https://www.youtube.com/watch?v=mntP0_EN-ZQ).
+```
+
+
+Control flow splitting (or short "splitting") is one of the techniques that the
+Certora Prover employs to speed up solving. In the remainder of this section, we
+will give an overview of how the technique works. This background should be
+helpful when using the settings described [here](control-flow-splitting-options)
+to prevent Prover timeouts.
+
+Splitting is best illustrated using the {term}`control flow graph` (CFG) of a given
+CVL rule.
+
+A single splitting step proceeds as follows:
+ - Pick a node with two successors in the CFG, the *split point*.
+ - Generate two new CFGs, we call them *splits*; both splits are copies of the 
+   original CFG, except that in the first (second) split, the edge to the first 
+   (second) successor has been removed. The algorithm also removes all nodes and 
+   edges that become unreachable through the initial edge removal.
+
+```{figure} split-step.png
+:name: single_split()
+:alt: A single splitting step
+:align: center
+:height: 300px
+
+Illustration of a single splitting step
+```
+
+
+There is an internal heuristic deciding which branching nodes to pick for each
+single splitting step.
+
+The following pseudo-code illustrates how Certora prover applies the single splitting 
+in a recursive fashion.
+
+```{code-block}
+:name: recursive splitting algorithm
+:caption: "Recursive splitting algorithm as pseudo code"
+
+Input: input_program_cfg
+
+worklist = []
+worklist.add([input_program_cfg, 0])
+
+while (worklist != [])
+    [cfg, current_depth] = worklist.pop()
+
+    res = smt_check(cfg, get_timeout_for(current_depth))
+    when (res) 
+        [SAT, model] -> return [SAT, model]
+        UNSAT -> continue
+        TIMEOUT -> 
+            if (current_depth == max_depth)
+                return timeout
+            else
+                [split_1, split_2] = split_single(cfg)
+                worklist.add([split_1, current_depth + 1])
+                worklist.add([split_2, current_depth + 1])
+return UNSAT
+```
+
+Intuitively, the algorithm explores the tree of all possible recursive
+splittings along a fixed sequence of split points up to the maximum splitting
+depth. We call the splits at maximum splitting depth *split leafs*. The
+exploration stops in either of the following three cases:
+ - if one split was found that is SAT (reasoning: if one split is SAT, then the
```

Comment:
I think `SAT` and `UNSAT` require some mental gymnastics.  I think something like "if a counterexample is found for any split, then the process stops" would be clearer.

Reply:
the problem is that it's not always a counterexample, sometimes it's a "witness example" ...
 -- so I chose SAT because it's the more "stable" concept here ...
I made a glossary entry explaining the background there.
 -- not needing the mental gymnastics is of course better 🤔 

I can try how it looks with "counterexample", and on first mention, or in a footnote, mention witness examples and SAT/UNSAT, pointing to the glossary ...

Reply:
hm, after revisiting this again, I feel like for this very technical section, SAT and UNSAT might be best, since they're just the most concise thing to put e.g. into the pseudo code ...
 -- there is a glossary entry for them, which also relates them to the concepts of counterexample and witness example.

I'll also add a sentence to the top of this section mentioning this relationship
 -- let me know if you think this won't do ...